### ISSUE TITLE
[2.2] Training Loop with Checkpointing + TensorBoard

### ISSUE THREAD

**Issue by poltergeistjoya** [poltergeistjoya/Vacant_Lot_Detection#13]
## Summary
Implement training loop with gradient accumulation, TensorBoard logging, learning rate scheduling, and checkpointing for interrupted jobs.

## Implementation Details
- **New file:** `vacant_lot/train.py`

### SegmentationTrainer Class
```python
class SegmentationTrainer:
    def train_epoch(self, epoch):
        '''One epoch: forward, backward, gradient accumulation'''
        
    def validate(self, epoch):
        '''Evaluate on validation set (Bronx patches)'''
        
    def save_checkpoint(self, epoch, metrics, is_best):
        '''Save model + optimizer state for resume'''
        
    def train(self, max_epochs=50):
        '''Main loop: train_epoch → validate → checkpoint'''
```

### Optimizer & Scheduler
- **Optimizer:** AdamW (lr=1e-4, weight_decay=1e-4)
  - Decoupled weight decay (Loshchilov & Hutter, 2019)
- **Scheduler:** CosineAnnealingLR (T_max=50, eta_min=1e-6)
  - Warm restarts improve convergence (Loshchilov & Hutter, 2017)
  - Default in SegFormer (Xie et al., 2021)

### Features
1. **Gradient Accumulation:** Simulate larger batch sizes on memory-limited GPU (Goyal et al., 2017)
2. **TensorBoard Logging:**
   - Train loss, val loss per epoch
   - IoU, Dice, Precision, Recall curves
   - Learning rate schedule
   - Histograms of weights
3. **Checkpointing:** Save best model + latest checkpoint
   - Allows resume on preemptible GPUs
4. **Early Stopping:** Stop if val metric plateaus for N epochs (patience=10)

### Output
- `modeling/outputs/models/nyc_2022/s3_{arch}_{encoder}_{run_id}.pt` — model weights
- `modeling/outputs/models/nyc_2022/s3_{arch}_{encoder}_{run_id}.json` — metadata (architecture, hyperparams, final metrics)
- `modeling/outputs/logs/nyc_2022/{run_id}/` — TensorBoard logs

## Acceptance Criteria
- [ ] Training loop runs without errors
- [ ] Loss decreases over epochs (sanity check)
- [ ] Validation metrics computed correctly
- [ ] Checkpoints saved and loadable
- [ ] TensorBoard logs generated and viewable
- [ ] Resume from checkpoint works (test on dummy data)
- [ ] Early stopping triggers correctly

## Dependencies
- torch
- tensorboard
- torch.utils.tensorboard.SummaryWriter

## References
- Loshchilov & Hutter (2019), AdamW, ICLR, arXiv:1711.05101
- Loshchilov & Hutter (2017), SGDR, ICLR, arXiv:1608.03983
- Xie et al. (2021), SegFormer, NeurIPS, arXiv:2105.15203
- Goyal et al. (2017), Accurate, Large Minibatch SGD, arXiv:1706.02677
